helix-core/src/case_conversion.rs

```diff
@@ -2,68 +2,302 @@ use crate::Tendril;
 
 // todo: should this be grapheme aware?
 
-pub fn to_pascal_case(text: impl Iterator<Item = char>) -> Tendril {
-    let mut res = Tendril::new();
-    to_pascal_case_with(text, &mut res);
-    res
+/// Converts each character into a different one, with zero context about surrounding characters
+pub fn simple_case_conversion(
+    text: impl Iterator<Item = char>,
+    buf: &mut Tendril,
+    transform_char: impl Fn(&char) -> char,
+) {
+    for c in text {
+        buf.push(transform_char(&c))
+    }
 }
 
-pub fn to_pascal_case_with(text: impl Iterator<Item = char>, buf: &mut Tendril) {
-    let mut at_word_start = true;
-    for c in text {
-        // we don't count _ as a word char here so case conversions work well
-        if !c.is_alphanumeric() {
-            at_word_start = true;
-            continue;
-        }
-        if at_word_start {
-            at_word_start = false;
-            buf.extend(c.to_uppercase());
+pub fn complex_case_conversion(
+    text: impl Iterator<Item = char>,
+    buf: &mut Tendril,
+    capitalize_first: bool,
+    separator: Option<char>,
+) {
+    let mut capitalize_next = capitalize_first;
+    let mut prev: Option<char> = None;
+
+    for c in text.skip_while(|ch| ch.is_whitespace()) {
+        if c.is_alphanumeric() {
+            if prev.is_some_and(|p| p.is_lowercase()) && c.is_uppercase() {
+                capitalize_next = true;
+            }
+            if capitalize_next {
+                buf.push(c.to_ascii_uppercase());
+                capitalize_next = false;
+            } else {
+                buf.extend(c.to_lowercase());
+            }
         } else {
-            buf.push(c)
+            capitalize_next = true;
+            if let Some(separator) = separator {
+                if prev.is_some_and(|p| p != separator) {
+                    buf.push(separator);
+                }
+            }
         }
```

Comment:
I would merge the conditions into one `if`, unless the compiler complains

Reply:
unfortunately:
```
`let` expressions in this position are unstable
see issue #53667 <https://github.com/rust-lang/rust/issues/53667> for more information
```